File: src/main/java/com/rfb/demo/rxjavatest/algorithm/leetCode3/maxCoins.java
package com.rfb.demo.rxjavatest.algorithm.leetCode3;

public class maxCoins {

    private int dfs(int[]nums, int left, int right){

        if(left == right){
            return nums[left];
        }

        if(left + 1 == right){
            if(nums[left] > nums[right]){
                return nums[left]*nums[right] + nums[left];
            }else{
                return nums[left]*nums[right] + nums[right];
            }
        }

        int temp = 0;
        for(int i = left+1; i < right; i++){

            int valueLeft = dfs(nums, left, i-1);
            int valueRight = dfs(nums, i+1, right);
            int value = nums[i-1]*nums[i+1]*nums[i];
            int all = valueLeft + valueRight + value;
            if(temp < all){
                temp = all;
            }
        }

        return temp;
    }

    public int maxCoins(int[] nums) {
        return dfs(nums, 0, nums.length-1);
    }

    public void test(){
        System.out.println(maxCoins(new int[]{3,1,5,8}));
    }

}

File: src/main/java/com/rfb/demo/rxjavatest/kotlin/leetcode/maxCoins.kt
package com.rfb.demo.rxjavatest.kotlin.leetcode

class maxCoins {

    private fun dfs(nums: IntArray, from:Int, to:Int):Int{

        if(from == to){
            val left = if(from == 0){1}else{nums[from-1]}
            val right = if(from == nums.size-1){1}else{nums[from+1]}
            return left * right * nums[from]
        }

        var max = 0
        for(i in from..to){

            val left = if(i == from){ 0 }else{ dfs(nums, from, i-1) }
            val right = if(i == to){ 0 }else{ dfs(nums, i+1, to)}
            val value = left + right + nums[i]
            if(value > max){
                max = value
            }

            if(from == 0 && to == 3){
                println("$i $max $value")
            }
        }

        return max
    }

    fun maxCoins(nums: IntArray): Int {
        return dfs(nums, 0, nums.size-1)
    }

    fun test(){
//        println(maxCoins(intArrayOf(3,1,5,8)))
        println(maxCoins(intArrayOf(3,1,5,8)))
    }
}